### ISSUE TITLE
Dead lock for multi-session

### ISSUE THREAD

**Issue by charles1129** [OP-TEE/optee_os#6226]
Dear Sirs,
> Recently we noticed a issue with our custom TA. If CA was killed using killall command and then restarted, sometimes it would stuck in uninterruptible sleep. It should not be a problem during normal work, but there might be a risk of CA being stuck on sleep after a kill by OOM killer or restarted for some other reason. 
It seems like there is some deadlock in TEE code causing this, so I conducted a stress test for hello world example.

### hello_world TA : 

- Add multi session flag in user_ta_header_defines.h:

```
#define TA_FLAGS (TA_FLAG_USER_MODE | TA_FLAG_EXEC_DDR | \
                 TA_FLAG_SINGLE_INSTANCE | TA_FLAG_MULTI_SESSION)
```

- Trigger panic in invoke:

```
    TEE_Result TA_InvokeCommandEntryPoint(void *sess_ctx __unused, uint32_t cmd_id,
                                      uint32_t param_types,
                                      TEE_Param params[4] __unused)
    {
        uint32_t exp_param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_NONE,
                                                   TEE_PARAM_TYPE_NONE,
                                                   TEE_PARAM_TYPE_NONE,
                                                   TEE_PARAM_TYPE_NONE);

        DMSG("%s(): begin called with cmd_id:%#" PRIx32 "\n", __func__, cmd_id);

        TEE_Panic(-1);
        ...
    }
```

### hello_world CA:
Execute hello_world CA in two while loop

    Process 1 run CA in while loop;
    Process 2 run CA in while loop;

### Debug:
After debug, I find it's often blocked in core/kernel/tee_ta_manager.c:
```
    static void tee_ta_unlink_session(struct tee_ta_session *s,
                        struct tee_ta_session_head *open_sessions)
    {
        mutex_lock(&tee_ta_mutex);

        ...

        while (s->ref_count != 1)
                 condvar_wait(&s->refc_cv, &tee_ta_mutex); // blocked here 

        TAILQ_REMOVE(open_sessions, s, link);

        mutex_unlock(&tee_ta_mutex);
    }

```

Do you have any better suggestions?

Best Regards

**Comment by jenswi-linaro** [comment]
It seems that the thread holding `tee_ta_mutex` either is killed by the OOM killer or perhaps is blocked by something else in the kernel driver.

**Comment by charles1129** [comment]
hi @jenswi-linaro 

I find that the deadlock reason is session ref_count != 1.

![image](https://github.com/OP-TEE/optee_os/assets/120549548/e240036f-fd92-4295-b96e-e7456c7e72e3)


so I think we should put session before close when panicked, right? 
```
diff --git a/core/kernel/tee_ta_manager.c b/core/kernel/tee_ta_manager.c
index b9fd3201e..9e2c84cbe 100644
--- a/core/kernel/tee_ta_manager.c
+++ b/core/kernel/tee_ta_manager.c
@@ -727,6 +727,7 @@ TEE_Result tee_ta_open_session(TEE_ErrorOrigin *err,

        if (!ctx || ctx->panicked) {
                DMSG("panicked, call tee_ta_close_session()");
+               tee_ta_put_session(s);
                tee_ta_close_session(s, open_sessions, KERN_IDENTITY);
                *err = TEE_ORIGIN_TEE;
                return TEE_ERROR_TARGET_DEAD;

```
Thanks

**Comment by charles1129** [comment]
hi @jenswi-linaro ,
I used your patch, and there has been no deadlock issue by now. But there is other issue: 
thread 2 destroyed ctx, but thread 1 still in **enter_invoke_cmd**, TEE will panic somewhere.

I try to fix it as:

```
diff --git a/core/kernel/tee_ta_manager.c b/core/kernel/tee_ta_manager.c
index b9fd3201e..0f5fd6879 100644
--- a/core/kernel/tee_ta_manager.c
+++ b/core/kernel/tee_ta_manager.c
@@ -793,14 +794,14 @@ TEE_Result tee_ta_invoke_command(TEE_ErrorOrigin *err,
        }

        ta_ctx = ts_to_ta_ctx(ts_ctx);
-       if (ta_ctx->panicked) {
-               DMSG("Panicked !");
-               destroy_ta_ctx_from_session(sess);
+       if (!ta_ctx) {
                *err = TEE_ORIGIN_TEE;
                return TEE_ERROR_TARGET_DEAD;
        }

        tee_ta_set_busy(ta_ctx);
+       if (ta_ctx->panicked)
+               goto out;

        sess->param = param;
        set_invoke_timeout(sess, cancel_req_to);
@@ -809,7 +810,9 @@ TEE_Result tee_ta_invoke_command(TEE_ErrorOrigin *err,
        sess->param = NULL;
        tee_ta_clear_busy(ta_ctx);

+out:
        if (ta_ctx->panicked) {
+               DMSG("Panicked !");
                destroy_ta_ctx_from_session(sess);
                *err = TEE_ORIGIN_TEE;
                return TEE_ERROR_TARGET_DEAD;
```

And fix the panic issues I found.
```
diff --git a/core/kernel/pseudo_ta.c b/core/kernel/pseudo_ta.c
index 0a85eda96..9defd0623 100644
--- a/core/kernel/pseudo_ta.c
+++ b/core/kernel/pseudo_ta.c
@@ -247,7 +247,7 @@ static const struct ts_ops pseudo_ta_ops = {

 bool is_pseudo_ta_ctx(struct ts_ctx *ctx)
 {
-       return ctx->ops == &pseudo_ta_ops;
+       return ctx && ctx->ops == &pseudo_ta_ops;
 }
```

```
diff --git a/core/include/kernel/user_mode_ctx.h b/core/include/kernel/user_mode_ctx.h
index 6d65d1470..2424808f4 100644
--- a/core/include/kernel/user_mode_ctx.h
+++ b/core/include/kernel/user_mode_ctx.h
@@ -25,8 +25,10 @@ static inline struct user_mode_ctx *to_user_mode_ctx(struct ts_ctx *ctx)
                return &to_user_ta_ctx(ctx)->uctx;
        else if (is_sp_ctx(ctx))
                return &to_sp_ctx(ctx)->uctx;
-       else
+       else if (is_stmm_ctx(ctx))
                return &to_stmm_ctx(ctx)->uctx;
+       else
+               panic("no such ctx");
 }

```

System will panic at here, I think this is as expected.
```
to_user_mode_ctx: panic("no such ctx");
ts_to_ta_ctx: panic("bad context");
```

How do you think of this?
By the way, do you think the same issue will appear in the open or close processes?

Thanks

**Comment by jenswi-linaro** [comment]
It's getting hard to follow exactly what you're proposing.
Would you mind creating a PR with the changes you think are needed at the moment?
We 

> By the way, do you think the same issue will appear in the open or close processes?

That can't be ruled out. Do you think it's feasible to create a couple of test cases for this to put in xtest? It's not clear to me exactly how you're testing this. I can help out with this if needed.

**Comment by jenswi-linaro** [comment]
I've been able to create a test case to reproduce the deadlock and then also the panic. I'm investigating this error further.

**Comment by charles1129** [comment]
ok, thank you.
I find one of the panic reason is:
```
Thread 1 trigger panic, and running in user mode or other steps,
Then thread 2 destroy the ctx.
Thread 1 still running  with invalid ctx and panicked.
```

![image](https://github.com/OP-TEE/optee_os/assets/120549548/3ff6b7a8-5f18-4fcb-912b-56bc3bfa6436)

**Comment by jenswi-linaro** [comment]
It seems the handling of panicked TAs has a few races. I have a good test case to reproduce it so I'm looking into this now.

**Comment by jenswi-linaro** [comment]
I believe I have a fix for this in https://github.com/OP-TEE/optee_os/pull/6281. Can you confirm that it fixes the issue?

**Comment by charles1129** [comment]
ok, I'll confirm, thank you.

**Comment by charles1129** [comment]
> I believe I have a fix for this in #6281. Can you confirm that it fixes the issue?

Test passed for v3.18, thank you for your help.

I'll try to apply the patch to TEE v3.8 or v2.4 next.

**Comment by jforissier** [comment]
@charles1129 if you want to give a `Tested-by: Your Name <your.name@your.email>` tag in the discussion for #6281 please feel free to do so and we will add it to the commit message. Thanks.

**Comment by charles1129** [comment]
Wentao Sun <wentao.sun@amlogic.com>
Thank you
